# Tanzania-Tilapia-ddRAD-analysis
This repository contains the scripts used to analyze ddRAD-seq data for Tanzanian tilapia populations to  exploring the genetic diversity, population structure, and hybridization in Nile tilapia (Oreochromis niloticus) and Rufiji tilapia (Oreochromis urolepis urolepis) populations. Analyses include Stacks de novo SNP discovery, PCA, DAPC, FST and hybridization detection using triangulaR.

## Overview of analyses

The analyses consist of four main steps:

1. De novo assembly and SNP calling using Stacks
2. Population-level filtering and summary statistics
3. Population genomic analyses (PCA, DAPC, FST)
4. Hybridization analysis using the R package TriangulaR

## Repository structure
├── README.md
└── scripts/
    ├── 01_denovo_stacks.sh # De novo assembly and SNP calling
    ├── 02_populations_stacks.sh # Population-level filtering and outputs
    ├── 03_Tilapia_population_genomics.R # PCA, DAPC, FST analyses
    └── 04_hybridization_triangulaR.R # Hybridization detection


## Software

- Stacks v2.68
- R v4.5.2
- TriangulaR (R package)

## Scripts

### 1. De novo assembly (Stacks)

`scripts/denovo_stacks.sh`

Runs `denovo_map.pl` with the parameters used in the study to assemble loci
and call SNPs from ddRAD-seq data.

### 2. Population-level analysis (Stacks)

`scripts/populations_stacks.sh`

Runs the `populations` module to filter SNPs and generate summary statistics
and output files (VCF, STRUCTURE, FST).

### 3. Tilapia_opulation genomics analysis (R)

`scripts/population_genomics.R`

Performs population genomic analyses based on SNP data generated by Stacks,
including principal component analysis (PCA), discriminant analysis of
principal components (DAPC), and estimation of population differentiation
(FST).

### 4. Hybridization analysis (R)

`scripts/hybridization_triangulaR.R`

Uses the R package **TriangulaR** to test for hybridization among populations
based on SNP data generated by Stacks.

## Notes
- Raw sequencing data are not included in this repository.

